File: 22_专题/RMQ问题/SparseTable.ts
/* eslint-disable no-inner-declarations */
/* eslint-disable no-param-reassign */

import assert from 'assert'

const INF = 2e15

/**
 * st表适用于可重复贡献问题/RMQ静态区间最值查询。
 * 可重复贡献问题是指对于运算`opt`，满足`x opt x = x`，则对应的区间询问就是一个可重复贡献问题。
 * !幂等性
 * 例如，最大值有 `max(a,a) = a`，gcd有 `gcd(a,a) = a` ，所以RMQ和区间GCD就是一个可重复贡献问题。
 * 像区间和就不具有这个性质，如果求区间和的时候采用的预处理区间重叠了，
 * 则会导致重曼部分被计算两次，这是我们所不愿意看到的。
 * 另外，opt还必须满足结合律才能使用ST表求解。
 * @see {@link https://oi-wiki.org/ds/sparse-table/}
 * @important
 * - cpu-cache optimized `dp[bit][n]`
 * - compress to 1d `dp[bit][n] -> dp[bit*n]`
 */
class SparseTable<S = number> {
  private readonly _e: () => S
  private readonly _op: (a: S, b: S) => S
  private readonly _dp: S[]
  private readonly _n: number

  constructor(nums: ArrayLike<S>, e: () => S, op: (a: S, b: S) => S) {
    const n = nums.length
    const size = 32 - Math.clz32(n)
    this._n = n
    this._e = e
    this._op = op
    this._dp = Array(size * n) // !dp[i][j]表示区间[j,j+2**i)的贡献值
    for (let i = 0; i < n; i++) this._dp[i] = nums[i]
    for (let i = 1; i < size; i++) {
      for (let j = 0; j < n - (1 << i) + 1; j++) {
        const pos = i * n + j
        this._dp[pos] = op(this._dp[pos - n], this._dp[pos - n + (1 << (i - 1))])
      }
    }
  }

  /**
   * 查询左闭右开区间`[start, end)`的贡献值.
   * 0 <= start <= end <= n.
   */
  query(start: number, end: number): S {
    if (start < 0) start = 0
    if (end > this._n) end = this._n
    if (start >= end) return this._e()
    const k = 31 - Math.clz32(end - start)
    return this._op(this._dp[k * this._n + start], this._dp[k * this._n + end - (1 << k)])
  }

  /**
   * 返回最大的`right`使得`[left,right)`内的值满足`check`.
   * 0 <= left <= right <= n.
   */
  maxRight(left: number, check: (s: S) => boolean): number {
    if (left === this._n) return this._n
    let ok = left
    let ng = this._n + 1
    while (ok + 1 < ng) {
      const mid = (ok + ng) >> 1
      if (check(this.query(left, mid))) {
        ok = mid
      } else {
        ng = mid
      }
    }
    return ok
  }

  /**
   * 返回最小的`left`使得`[left,right)`内的值满足`check`.
   * 0 <= left <= right <= n.
   */
  minLeft(right: number, check: (s: S) => boolean): number {
    if (!right) return 0
    let ok = right
    let ng = -1
    while (ng + 1 < ok) {
      const mid = (ok + ng) >> 1
      if (check(this.query(mid, right))) {
        ok = mid
      } else {
        ng = mid
      }
    }
    return ok
  }
}

/**
 * 元素类型为`int32`类型的st表.
 */
class SparseTableInt32 {
  private readonly _e: () => number
  private readonly _op: (a: number, b: number) => number
  private readonly _dp: Int32Array
  private readonly _n: number

  constructor(nums: ArrayLike<number>, e: () => number, op: (a: number, b: number) => number) {
    const n = nums.length
    const size = 32 - Math.clz32(n)
    this._n = n
    this._e = e
    this._op = op
    this._dp = new Int32Array(size * n) // !dp[i][j]表示区间[j,j+2**i)的贡献值
    for (let i = 0; i < n; i++) this._dp[i] = nums[i]
    for (let i = 1; i < size; i++) {
      for (let j = 0; j < n - (1 << i) + 1; j++) {
        const pos = i * n + j
        this._dp[pos] = op(this._dp[pos - n], this._dp[pos - n + (1 << (i - 1))])
      }
    }
  }

  /**
   * 查询左闭右开区间`[start, end)`的贡献值.
   * 0 <= start <= end <= n.
   */
  query(start: number, end: number): number {
    if (start < 0) start = 0
    if (end > this._n) end = this._n
    if (start >= end) return this._e()
    const k = 31 - Math.clz32(end - start)
    return this._op(this._dp[k * this._n + start], this._dp[k * this._n + end - (1 << k)])
  }

  /**
   * 返回最大的`right`使得`[left,right)`内的值满足`check`.
   * 0 <= left <= right <= n.
   */
  maxRight(left: number, check: (s: number) => boolean): number {
    if (left === this._n) return this._n
    let ok = left
    let ng = this._n + 1
    while (ok + 1 < ng) {
      const mid = (ok + ng) >> 1
      if (check(this.query(left, mid))) {
        ok = mid
      } else {
        ng = mid
      }
    }
    return ok
  }

  /**
   * 返回最小的`left`使得`[left,right)`内的值满足`check`.
   * 0 <= left <= right <= n.
   */
  minLeft(right: number, check: (s: number) => boolean): number {
    if (!right) return 0
    let ok = right
    let ng = -1
    while (ng + 1 < ok) {
      const mid = (ok + ng) >> 1
      if (check(this.query(mid, right))) {
        ok = mid
      } else {
        ng = mid
      }
    }
    return ok
  }
}

export { SparseTable, SparseTableInt32 }

if (require.main === module) {
  const st = new SparseTable(
    [9, 12, 3, 7, 15],
    () => 0,
    (a, b) => Math.max(a, b)
  )
  assert.strictEqual(st.query(0, 1), 9)
  assert.strictEqual(st.query(0, 3), 12)
  assert.strictEqual(st.query(0, 5), 15)
  assert.strictEqual(st.query(1, 1), 0)
  console.log(st.maxRight(0, preMax => preMax <= 12))
}

File: 22_专题/RMQ问题/SparseTable.go
// st表, 查询区间最大值以及对应的下标(多个最大值时取最小的下标).

package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
)

const INF32 int32 = 1 << 30

// https://judge.yosupo.jp/problem/staticrmq
func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var n, q int32
	fmt.Fscan(in, &n, &q)
	nums := make([]int32, n)
	for i := int32(0); i < n; i++ {
		fmt.Fscan(in, &nums[i])
	}
	e := func() S { return S{min: INF32, index: -1} }
	op := func(s1, s2 S) S {
		if s1.min < s2.min {
			return s1
		}
		if s1.min > s2.min {
			return s2
		}
		return S{min: s1.min, index: min32(s1.index, s2.index)}
	}
	rmq := NewSparseTableFast(n, func(i int32) S { return S{min: nums[i], index: i} }, e, op)

	for i := int32(0); i < q; i++ {
		var start, end int32
		fmt.Fscan(in, &start, &end)
		fmt.Fprintln(out, rmq.Query(start, end).min)
	}
}

// RangeMaxWIthIndex

type S = struct {
	min   int32
	index int32
}

// Static RMQ, O(n)预处理, O(1)查询.
type SparseTableFast struct {
	n        int32
	leaves   []S
	pre, suf []S
	st       *SparseTable
	data     []int32
	e        func() S
	op       func(S, S) S
}

func NewSparseTableFast(n int32, f func(int32) S, e func() S, op func(S, S) S) *SparseTableFast {
	res := &SparseTableFast{}
	bNum := n >> 4
	leaves := make([]S, n)
	for i := int32(0); i < n; i++ {
		leaves[i] = f(i)
	}
	pre, suf := append(leaves[:0:0], leaves...), append(leaves[:0:0], leaves...)
	for i := int32(1); i < n; i++ {
		if i&15 != 0 {
			pre[i] = op(pre[i-1], leaves[i])
		}
	}
	for i := n - 1; i > 0; i-- {
		if i&15 != 0 {
			suf[i-1] = op(leaves[i-1], suf[i])
		}
	}
	st := NewSparseTable(bNum, func(i int32) S { return suf[i<<4] }, e, op)

	// 处理长度小于或等于16的查询
	// 在区间 [i, i+16) 内，如果 i+j 的位置上的值是 [i, i+j] 这个子区间的最小值，那么就将 j-th 位设置为1
	data := make([]int32, n)
	stack := int32(0)
	for i := n - 1; i >= 0; i-- {
		stack = (stack << 1) & 65535
		for stack > 0 {
			k := lowbit(stack)
			tmp := op(leaves[i], leaves[i+k])
			if tmp != leaves[i] {
				break
			}
			stack &= ^(1 << k)
		}
		stack |= 1
		data[i] = stack
	}
	res.n = n
	res.leaves = leaves
	res.pre, res.suf = pre, suf
	res.st = st
	res.data = data
	res.e = e
	res.op = op
	return res
}

func NewSparseTableFastFrom(leaves []S, e func() S, op func(S, S) S) *SparseTableFast {
	return NewSparseTableFast(int32(len(leaves)), func(i int32) S { return leaves[i] }, e, op)
}

func (st *SparseTableFast) Query(start, end int32) S {
	if start >= end {
		return st.e()
	}
	if end-start <= 16 {
		d := st.data[start] & ((1 << (end - start)) - 1)
		return st.leaves[start+topbit(d)]
	}
	end--
	a, b := start>>4, end>>4
	x := st.st.Query(a+1, b)
	x = st.op(st.suf[start], x)
	x = st.op(x, st.pre[end])
	return x
}

// 返回最大的 right 使得 [left,right) 内的值满足 check.
func (ds *SparseTableFast) MaxRight(left int32, check func(e S) bool) int32 {
	if left == ds.n {
		return ds.n
	}
	ok, ng := left, ds.n+1
	for ok+1 < ng {
		mid := (ok + ng) >> 1
		if check(ds.Query(left, mid)) {
			ok = mid
		} else {
			ng = mid
		}
	}
	return ok
}

// 返回最小的 left 使得 [left,right) 内的值满足 check.
func (ds *SparseTableFast) MinLeft(right int32, check func(e S) bool) int32 {
	if right == 0 {
		return 0
	}
	ok, ng := right, int32(-1)
	for ng+1 < ok {
		mid := (ok + ng) >> 1
		if check(ds.Query(mid, right)) {
			ok = mid
		} else {
			ng = mid
		}
	}
	return ok
}

// SparseTable 稀疏表: st[j][i] 表示区间 [i, i+2^j) 的贡献值.
type SparseTable struct {
	st [][]S
	e  func() S
	op func(S, S) S
	n  int32
}

func NewSparseTable(n int32, f func(int32) S, e func() S, op func(S, S) S) *SparseTable {
	res := &SparseTable{}

	b := bits.Len32(uint32(n))
	st := make([][]S, b)
	for i := range st {
		st[i] = make([]S, n)
	}
	for i := int32(0); i < n; i++ {
		st[0][i] = f(i)
	}
	for i := 1; i < b; i++ {
		for j := int32(0); j+(1<<i) <= n; j++ {
			st[i][j] = op(st[i-1][j], st[i-1][j+(1<<(i-1))])
		}
	}
	res.st = st
	res.e = e
	res.op = op
	res.n = n
	return res
}

func NewSparseTableFrom(leaves []S, e func() S, op func(S, S) S) *SparseTable {
	return NewSparseTable(int32(len(leaves)), func(i int32) S { return leaves[i] }, e, op)
}

// 查询区间 [start, end) 的贡献值.
func (st *SparseTable) Query(start, end int32) S {
	if start >= end {
		return st.e()
	}
	b := bits.Len(uint(end-start)) - 1 // log2
	return st.op(st.st[b][start], st.st[b][end-(1<<b)])
}

// 返回最大的 right 使得 [left,right) 内的值满足 check.
func (ds *SparseTable) MaxRight(left int32, check func(e S) bool) int32 {
	if left == ds.n {
		return ds.n
	}
	ok, ng := left, ds.n+1
	for ok+1 < ng {
		mid := (ok + ng) >> 1
		if check(ds.Query(left, mid)) {
			ok = mid
		} else {
			ng = mid
		}
	}
	return ok
}

// 返回最小的 left 使得 [left,right) 内的值满足 check.
func (ds *SparseTable) MinLeft(right int32, check func(e S) bool) int32 {
	if right == 0 {
		return 0
	}
	ok, ng := right, int32(-1)
	for ng+1 < ok {
		mid := (ok + ng) >> 1
		if check(ds.Query(mid, right)) {
			ok = mid
		} else {
			ng = mid
		}
	}
	return ok
}

func topbit(x int32) int32 {
	if x == 0 {
		return -1
	}
	return int32(bits.Len32(uint32(x)) - 1)
}

func lowbit(x int32) int32 {
	if x == 0 {
		return -1
	}
	return int32(bits.TrailingZeros32(uint32(x)))
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func min32(a, b int32) int32 {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func max32(a, b int32) int32 {
	if a > b {
		return a
	}
	return b
}
